pyqtgraph/graphicsItems/ImageItem.py

```diff
@@ -242,6 +242,9 @@ def setImage(self, image=None, autoLevels=None, **kargs):
         levels             (min, max) The minimum and maximum values to use when rescaling the image
                            data. By default, this will be set to the minimum and maximum values
                            in the image. If the image array has dtype uint8, no rescaling is necessary.
+        levelSamples       (default: 65536) When determining minimum and maximum values, ImageItem
+                           only inspects a subset of pixels no larger than this number.
```

Comment:
Very good point.
I did not think about the 3d stacks of images also supported by this. I also think that the number should be per image slice.

Could you specify what the default behavior should be?
Right now it samples 65536 points across the entire set, and not even randomly, but in the same grid for each slice. I cannot see that as ideal for a deep stack of images. I am also pretty sure that a stack more than 65536 deep will hang the code now.
So I am in favor of changing the default behavior (for stacked images only).
Is 65536 samples per slice(=single image) okay?


Please explain why "grayscale" comes up here.... Does this even run for multichannel images? I thought the levels are specifically used in color-mapped images.

Reply:
As far as I know, `ImageItem` is only meant to work with a single image, so it's either a 2D array with shape `(px, py)` or 3D with shape `(px, py, ch)` where `ch` can be 3 for RGB or 4 for RGBA. Maybe you're thinking of `ImageView`, which does take stacks of images.

Yeah "grayscale" wasn't precise, I was just referring to monochrome as opposed to RGB(A).

Reply:
I really like `while img.shape[0]*img.shape[1] > level_samples` as the loop condition. It requires no additional checks for array dimensions, and no separation of code paths for the 2D arrays of a single channel image and the 3D arrays of RGB / RGBA images.

This *does* change the default behaviour now to perform the min/max search over 65536 * 3 values for RGB images, and 65536 * 4 values for RGBA. But the previous behaviour of sampling less pixels after adding an alpha channel... that doesn't make much sense to me.

This will have some impact on performance, but the user *can* now make it run even faster than before if they are willing to reduce the levelSamples further.

What do you think, @ixjlyons ?

Reply:
Sorry for the confusion of 2D/3D images. All my applications for this are single channel :)
Thank you for setting this straight.